books/kestrel/c/syntax/types.lisp

```diff
@@ -771,24 +908,71 @@
      @(':void') is compatible with @(':unknown'),
      as is @(':bool'),
      but @(':void') is <i>not</i> compatible with @(':bool')."))
-  (b* ((x (type-fix x))
-       (y (type-fix y)))
-    (or (equal x y)
-        (type-case x :unknown)
-        (type-case y :unknown)
-        (and (type-integerp x) (type-case y :enum))
-        (and (type-case x :enum) (type-integerp y))))
+  (or (type-case y :unknown)
+      (type-case
+        x
+        :unknown t
+        :pointer (type-case
+                   y
+                   :pointer (type-compatiblep x.to y.to)
+                   :otherwise nil)
+        :otherwise (or (equal (type-fix x) (type-fix y))
+                       (and (type-integerp x) (type-case y :enum))
+                       (and (type-case x :enum) (type-integerp y)))))
+  :measure (type-count x)
   :hooks (:fix)
 
   ///
 
   (defrule type-compatiblep-reflexive
     (type-compatiblep x x)
-    :enable type-compatiblep)
+    :induct t)
 
   (defrule type-compatiblep-symmetric
     (equal (type-compatiblep y x)
            (type-compatiblep x y))
+    :induct t))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(define type-composite ((x typep) (y typep))
```

Comment:
Should the composite of unknown and not-unknown be unknown instead? Perhaps it makes no difference in our current type system. In general, are there situations in which, in correct code, unknown stands for an actual type T that, when composited with a not-unknown T', should give something slightly different from T'? Maybe it needs a more refined type system for this to happen?

Reply:
Yes I think our type system would need to be further refined for such a case to arise (specifically, we'd need array sizes or full function prototypes). As of now, compositing the types will never change either type other than refining the unknowns. It can effect which of the two types is returned though. For instance, `(type-composite (type-enum) (type-sint))` would currently return `(type-enum)` (in arbitrary choice between the two). But if we abstract the enum, we get `(type-composite (type-unknown) (type-sint))` which evaluates to `(type-sint)`. That is fine though. Both answers are acceptable composite types for the two original types (the only constraint in this case is that the composite is compatible with both input types).

Reply:
Yes, `(type-composite (type-enum) (type-unknown))` would yield `(type-enum)`. But that seems fine, no? We consider enums compatible with integers, and the only requirement of composite types outside of arrays and functions is that the constructed type is compatible with both of the two input types.

Reply:
I guess so. Looking at [C17:6.2.7], it's not clear to me that it defines precisely what the composite type of two types is. I think it's fine to leave the code as you have it there.